dbt/adapters/redshift/connections.py

```diff
@@ -174,60 +197,59 @@ def get_connect_method(self):
-            # this requirement is really annoying to encode into json schema,
-            # so validate it here
-            if self.credentials.password is None:
-                raise FailedToConnectError(
-                    "'password' field is required for 'database' credentials"
-                )
-
-            def connect():
-                logger.debug("Connecting to redshift with username/password based auth...")
-                c = redshift_connector.connect(
-                    user=self.credentials.user,
-                    password=self.credentials.password,
-                    **kwargs,
-                )
-                if self.credentials.autocommit:
-                    c.autocommit = True
-                if self.credentials.role:
-                    c.cursor().execute("set role {}".format(self.credentials.role))
-                return c
+    @property
+    def _database_kwargs(self):
+        if self.credentials.password is None:
+            raise FailedToConnectError("'password' field is required for 'database' credentials")
+
+        logger.debug("Connecting to redshift with username/password based auth...")
+        return {
+            "user": self.credentials.user,
+            "password": self.credentials.password,
+        }
 
-        elif method == RedshiftConnectionMethod.IAM:
-            if not self.credentials.cluster_id and "serverless" not in self.credentials.host:
-                raise FailedToConnectError(
-                    "Failed to use IAM method. 'cluster_id' must be provided for provisioned cluster. "
-                    "'host' must be provided for serverless endpoint."
-                )
-
-            def connect():
-                logger.debug("Connecting to redshift with IAM based auth...")
-                c = redshift_connector.connect(
-                    iam=True,
-                    db_user=self.credentials.user,
-                    password="",
-                    user="",
-                    cluster_identifier=self.credentials.cluster_id,
-                    profile=self.credentials.iam_profile,
-                    **kwargs,
-                )
-                if self.credentials.autocommit:
-                    c.autocommit = True
-                if self.credentials.role:
-                    c.cursor().execute("set role {}".format(self.credentials.role))
-                return c
+    @property
+    def _iam_user_kwargs(self):
+        iam_kwargs = self._iam_kwargs
+
+        if self.credentials.access_key_id and self.credentials.secret_access_key:
+            iam_kwargs.update(
+                {
+                    "access_key_id": self.credentials.access_key_id,
+                    "secret_access_key": self.credentials.secret_access_key,
+                }
+            )
 
         else:
-            raise FailedToConnectError("Invalid 'method' in profile: '{}'".format(method))
+            iam_kwargs.update(
```

Comment:
The original logic flow allows for `iam_profile` to be empty and pass successfully. There's a unit test for it. So I check if `access_key_id` and `secret_access_key` are both provided, otherwise default to `iam_profile`. I'm not checking for the scenario where exactly one of `access_key_id` or `secret_access_key` are provided. This would lead to the user providing one of those, and then it tries to connect via the "profile, but no profile provided" method.